forte/data/data_store.py

```diff
@@ -382,22 +398,63 @@ def delete_entry(self, tid: int):
         Args:
             tid (int): Unique id of the entry.
         """
-        # We call ``get_entry()`` to obtain the entry, its `type_id`, and the
-        # index in the ``entry_type`` list (`index_id`). We remove the entry from
-        # `__entry_dict` when its information is retrieved. Then, the `type_id`
-        # and the ``index_id`` are passed to `_delete_entry_by_loc`.
-        raise NotImplementedError
-
-    def _delete_entry_by_loc(self, type_id: int, index_id: int):
-        r"""It removes an entry of ``index_id`` by taking both the `type_id`
-        and ``index_id``. Called by ``delete_entry()``.
+        try:
+            # get `entry data` and remove it from entry_dict
+            entry_data = self.__entry_dict.pop(tid)
+        except KeyError as e:
+            raise KeyError(
+                f"The specified tid [{tid}] "
+                f"does not correspond to an existing entry data "
+            ) from e
+
+        _, _, tid, type_name = entry_data[:4]
+        try:
+            target_list = self.__elements[type_name]
+        except KeyError as e:
+            raise KeyError(
+                f"The specified type [{type_name}] "
+                f"does not exist in self.__elements."
+            ) from e
+        # complexity: O(lgn)
+        # if it's annotation type, use bisect to find the index
+        if self._is_annotation(type_name):
+            entry_index = bisect_left(target_list, entry_data)
+        else:  # if it's group or link, use the index in entry_list
+            entry_index = entry_data[-1]
+
+        if (
+            entry_index >= len(target_list)
+            or target_list[entry_index] != entry_data
+        ):
+            raise ValueError(
+                f"Entry [{entry_data[:4]}] is not found" f"in the targetlist"
+            )
+
+        self._delete_entry_by_loc(type_name, entry_index)
+
+    def _delete_entry_by_loc(self, type_name: str, index_id: int):
+        r"""It removes an entry of `index_id` by taking both the `type_id`
+        and `index_id`. Called by `delete_entry()`.
+        This function will raise an IndexError if the `type_id` or `index_id`
+        is invalid.
 
         Args:
             type_id (int): The index of the list in ``self.__elements``.
             index_id (int): The index of the entry in the list.
         """
-        # We then remove the entry data from the `type_id`th list.
-        raise NotImplementedError
+        try:
+            target_list = self.__elements[type_name]
+        except KeyError as e:
+            raise KeyError(
+                f"The specified type [{type_name}] "
+                f"does not exist in self.__elements."
```

Comment:
`self.__elements` is not a good hint for exception, should better explain the situation here.